### PR TITLE
 Replace use of deprecated rfc822 module with email package

### PR DESCRIPTION
**Pull request by rodrigc**
Moved from: https://github.com/twisted/twisted/pull/162

See:
https://twistedmatrix.com/trac/ticket/7868

### CODE REVIEW

File: docs/core/examples/courier.py [hunk 1]
```diff
@@ -92,7 +92,8 @@ def filterMessage(self):
         A trivial example is included.
         """
         try:
-            m = Message(open(self.messageFilename))
+            emailParser = email.parser.Parser()
+            emailParser.parse(open(self.messageFilename))
```
Comment (markrwilliams) [inline docs/core/examples/courier.py]: This appears to leak the file;  [`email.parser.Parser.parse` completely consumes the file's contents but doesn't close it.](https://github.com/python/cpython/blob/master/Lib/email/parser.py#L42-L58).  If my reading of that is correct, this should be wrapped in an with statement:

``` python
with open(self.messageFilename) as f:
    emailParser.parse(f)
```

I realize this isn't part of a test, and isn't anything other than a stubbed-out example, but it'd be nice if our examples encouraged good resource management.

File: twisted/mail/test/test_imap.py [hunk 2]
```diff
@@ -1871,6 +1871,45 @@ def test_searchNot(self):
                 (None, None)))
 
 
+    def test_searchBefore(self):
+        """
```
Comment (markrwilliams) [inline twisted/mail/test/test_imap.py:1875]: Thanks for these tests!

**Comment by markrwilliams** [comment]
@rodrigc Looks good.  If/when the tests pass please merge this.

**EDIT** Though I would like to see the example changed I think that might be out of scope for this PR.  Fixing it now is up to you.  I'll watch the outcome and open a new ticket for it if necessary.

**EDIT** (again).  Looks like @rodrigc is once again faster than greased lightning.  Merge when the builds pass!

File: twisted/mail/test/test_imap.py [hunk 3]
```diff
@@ -1871,6 +1871,45 @@ def test_searchNot(self):
...
+        self.assertFalse(
+            self.server.search_BEFORE(self.earlierQuery, self.seq, self.msg))
+        self.assertFalse(
+            self.server.search_BEFORE(self.sameDateQuery, self.seq, self.msg))
+        self.assertTrue(
+            self.server.search_BEFORE(self.laterQuery, self.seq, self.msg))
+
+
+    def test_searchOn(self):
```
Comment (adiroiban) [inline twisted/mail/test/test_imap.py:1887]: @markrwilliams  does this test looks valid to you?

**Pull request merged**